ComputerCraft Integration for Eureka Ship Helm

TechTastic opened the pull request:
Turns ShipHelmBlock into a CC Peripheral

Added ShipHelmPeripheral and ShipHelmPeripheralProvider
Minor change to EurekaModFabric and EurekaModForge to test for if ComputerCraft is loaded and/or disabled via configs
Added Config Option to disable ComputerCraft integration
Added CC Tweaked and CC Restitched to build.gradle for Forge and Fabric respectively using Curse Maven and modCompileOnly

ShipHelmPeripheral provides multiple methods to manipulate ad read information to/from Eureka ships
 - forward() 
   - changes forwardImpulse (1.0f) in the SeatedControllingPlayer attachment and holds it for a second
 - reverse()
    - changes forwardImpulse (-1.0f) in the SeatedControllingPlayer attachment and holds it for a second
 - turnLeft()
    - changes leftImpulse (1.0f) in the SeatedControllingPlayer attachment and holds it for a second
 - turnRight()
    - changes leftImpulse (-1.0f) in the SeatedControllingPlayer attachment and holds it for a second
 - raise()
    - changes upImpulse (1.0f) in the SeatedControllingPlayer attachment and holds it for a second
 - lower()
    - changes upImpulse (-1.0f) in the SeatedControllingPlayer attachment and holds it for a second
 - isCruising()
    - grabs the boolean in SeatedControllingPlayer
 - startCruising()
   - changes the boolean in SeatedControllingPlayer
 - stopCruising()
   - changes the boolean in SeatedControllingPlayer
 - isAligning()
   - grabs the boolean in SeatedControllingPlayer
 - startAlignment()
   - changes the boolean in SeatedControllingPlayer
 - stopAlignment()
   - changes the boolean in SeatedControllingPlayer
 - isAligned()
   - checks the current omega.y from ShipData to the directional facing of the Ship Helm
 - canDisassemble()
   - grabs the boolean from EurekaShipControl
 - isDisassembling()
   - grabs the boolean from EurekaShipControl
 - disassemble()
   - calls to the disassemble() method in the ShipHelmBlockEntity
 - assemble()
   - calls to the disassemble() method in the ShipHelmBlockEntity
 - getBalloonAmount()
   - grabs the int from EurekaShipControl
 - getAnchorAmount()
   - grabs the int from EurekaShipControl
 - getActiveAnchorAmount()
   - grabs the int from EurekaShipControl
 - areAnchorsActive()
   - grabs the int from EurekaShipControl
 - getShipHelmAmount()
   - grabs the int from EurekaShipControl
 - getShipName()
   - calls to getName() from ShipData
 - setShipName()
   - calls to setName() from ShipData
 - getShipID()
   - calls to getId() from ShipData
 - getMass()
   - calls to getMass() from ShipInertiaData
 - getVelocity()
   - calls to getVelocity() from ShipData
 - getPosition()
   - calls to getShipPositionInWorldCoordinates() from ShipTransform
 - getScale()
   - calls to getShipPositionInWorldCoordinatesScaling() from ShipTransform

TechTastic committed:
Initial Commit

CC:R Fabric
------------------------------------------------
Added ShipHelmPeripheral and ShipHelmPeripheralProvider
Added a Test for ComputerCraft being loaded
Added forward(), reverse(), turnLeft(), turnRight(), raise(), and lower() to control Eureka ships via computers

TechTastic committed:
CC:R Fabric
------------------------------------------------
Removed getName() method
Added the following methods:
 - isCruising()
 - startCruising()
 - stopCruising()
 - isAligning()
 - startAlignment()
 - stopAlignment()
 - isAligned()
 - canDisassemble()
 - isDisassembling()
 - disassemble()
 - assemble()
 - getBalloonAmount()
 - getAnchorAmount()
 - getActiveAnchorAmount()
 - areAnchorsActive()
 - getShipHelmAmount()
 - getShipName()
 - setShipName()
 - getShipID()
 - getMass()
 - getVelocity()
 - getPosition()
 - getScale()

CC:T Forge
------------------------------------------------
Added ShipHelmPeripheral and ShipHelmPeripheralProvider
Added a Test for ComputerCraft being loaded
Added the following methods:
 - forward()
 - reverse()
 - turnLeft()
 - turnRight()
 - raise()
 - lower()
 - isCruising()
 - startCruising()
 - stopCruising()
 - isAligning()
 - startAlignment()
 - stopAlignment()
 - isAligned()
 - canDisassemble()
 - isDisassembling()
 - disassemble()
 - assemble()
 - getBalloonAmount()
 - getAnchorAmount()
 - getActiveAnchorAmount()
 - areAnchorsActive()
 - getShipHelmAmount()
 - getShipName()
 - setShipName()
 - getShipID()
 - getMass()
 - getVelocity()
 - getPosition()
 - getScale()

copperwarrior commented:
canDisassemble, isDisassembling, and isAligning should not be revealed to the client because they operate in ways that a client may find confusing.

TechTastic commented:
> canDisassemble, isDisassembling, and isAligning should not be revealed to the client because they operate in ways that a client may find confusing.

I both agree and disagree

They might find it confusing but it can also be a great boon for intricate ship control programs

Such as that you dont move when you are aligning, so the smart developer knowing that can use isAligning() to check before attempting any movement and can take appropriate action or a person may want to be able to disassemble() but may want to simply check if they even can, thus theyll use canDisassemble() and depoending on the outcome, take appropriate action

copperwarrior commented:
The smart developer can also keep track of when they start and stop aligning, and detect if a ship is able to disassemble at all by using the angles.
If you intend to keep these methods, they need to work in a manner consistent with what the client would expect.

ewoudje commented:
I have to agree with copper, although i think canDissassemble is fine, but i think its better to hide the isAligning and isDissambling mainly cus these could be replaced at any time. (disassembly is still not a working thing so). Maybe also expose a getRotation

ewoudje commented on fabric/src/main/java/org/valkyrienskies/eureka/fabric/integrations/cc_restitched/ShipHelmPeripheral.java:
```diff
@@ -0,0 +1,572 @@
+
+public class ShipHelmPeripheral implements IPeripheral {
+    Level world;
+    BlockPos pos;
```
Make these fields private since they're not used outside the class itself.

TechTastic replied:
alright

ewoudje commented on fabric/src/main/java/org/valkyrienskies/eureka/fabric/integrations/cc_restitched/ShipHelmPeripheral.java:
```diff
@@ -0,0 +1,572 @@
+                if (control != null) {
+                    SeatedControllingPlayer fakePlayer = ship.getAttachment(SeatedControllingPlayer.class);
+                    if (fakePlayer == null) {
+                        fakePlayer = new SeatedControllingPlayer(world.getBlockState(pos).getValue(BlockStateProperties.HORIZONTAL_FACING).getOpposite());
```
Is this pos you are using the helm position or the computer position? I would like it to use the orientation of the helm for this.

TechTastic replied:
the Helm as the position is passed from the PeripheralProvider which gives the Peripheral's position (which is the Ship Helm)

ewoudje commented on fabric/src/main/java/org/valkyrienskies/eureka/fabric/integrations/cc_restitched/ShipHelmPeripheral.java:
```diff
@@ -0,0 +1,572 @@
+        return new Object[0];
+    }
+
+    public boolean applyThrust(String direction) throws LuaException {
```
Can you add some comments here and there cus this is a very big method wich is fine, but its easier for later maintaining that there are some comments.

TechTastic replied:
sure, its basically me forcing the impulse for whatever direction to stay at 1.0f for 20 game ticks (~1 second)

ewoudje commented on fabric/src/main/java/org/valkyrienskies/eureka/fabric/integrations/cc_restitched/ShipHelmPeripheral.java:
```diff
@@ -0,0 +1,572 @@
+
+    @LuaFunction
+    public final int getAnchorAmount() throws LuaException {
+        if (!world.isClientSide()) {
```
So im unsure if you need world.isClientSide everywhere? is this not only called server-side. Also i recommend nesting everything less, so more in the style of `if (world.isClientSide()) return 0;` and then more code behind it without any nesting. This is easier to read.

ewoudje replied:
(this is ofc not a requirement for merging but just a tip)

TechTastic replied:
fair, i wasnt thinking in that vein, ill try it

TechTastic commented:
Ok, so Fabric is fully working and tested but Forge can not be tested in development due to CC: Tweaked either not remapping despite modCompileOnly or refusing to load and I can not outside of the IDE due to CentOS and Forge disagreeing on where .minecraft is and whether there is a launcher_profiles.json

TechTastic committed:
Final Commit

CC:T Forge
------------------------------------------------
Added EurekaPeripheralProviders to prevent NoClassFound exception of registering new PeripheralProviders

TechTastic commented:
This is now fully working

Game does not crash nor error out due to missing classes if CCT/CCR are not found and Forge matches Fabric


Both Modloader Builds have been tested outside the dev environment via CurseForge

nconder committed:
Fabric is fully working, Forge is untested due to CC:T refusing to run in dev

CC:R Fabric
------------------------
Added EurekaPeripheralPrvider to prevent ClassNotFoundError
Removed isAligning(), canDisassemble(), isDisassembling(), and isAligned()
Added getRotation()

CC:T Forge
------------------------
Removed isAligning(), canDisassemble(), isDisassembling(), and isAligned()
Added getRotation()

ewoudje reviewed (approved):
Looking good
